contributingGuides/TS_STYLE.md

```diff
@@ -0,0 +1,515 @@
+
+  // Good
+  const name = user?.name ?? "default name";
+  ```
+
+<a name="type-inference"></a><a name="1.9"></a>
+
+- [1.9](#type-inference) **Type Inference**: When possible, allow the compiler to infer type of variables.
+
+  ```ts
+  // Bad
+  const foo: string = "foo";
+  const [counter, setCounter] = useState<number>(0);
+
+  // Good
+  const foo = "foo";
+  const [counter, setCounter] = useState(0);
+  const [username, setUsername] = useState<string | undefined>(undefined); // Username is a union type of string and undefined, and its type cannot be interred from the default value of undefined
+  ```
+
+  For function return types, default to always typing them unless a function is simple enough to reason about its return type.
+
+  > Why? Explicit return type helps catch errors when implementation of the function changes. It also makes it easy to read code even when TypeScript intellisense is not provided.
+
+  ```ts
+  function simpleFunction(name: string) {
+    return `hello, ${name}`;
+  }
+
+  function complicatedFunction(name: string): boolean {
+    // ... some complex logic here ...
+    return foo;
+  }
+  ```
+
+<a name="jsdoc"></a><a name="1.10"></a>
+
+- [1.10](#jsdoc) **JSDoc**: Omit comments that are redundant with TypeScript. Do not declare types in `@param` or `@return` blocks. Do not write `@implements`, `@enum`, `@private`, `@override`. eslint: [`jsdoc/no-types`](https://github.com/gajus/eslint-plugin-jsdoc/blob/main/.README/rules/no-types.md)
+
+  ```ts
+  // bad
+  /**
+   * @param {number} age
+   * @returns {boolean} Whether the person is a legal drinking age or nots
+   */
+  function canDrink(age: number): boolean {
+    return age >= 21;
+  }
+
+  // good
+  /**
+   * @param age
+   * @returns Whether the person is a legal drinking age or nots
+   */
+  ```
+
+<a name="proptypes-and-defaultprops"></a><a name="1.11"></a>
+
+- [1.11](#proptypes-and-defaultprops) **`propTypes` and `defaultProps`**: Do not use them. Use object destructing to assign default values if necessary.
+
+  > Refer to [the propTypes Migration Table](./PROPTYPES_CONVERSION_TABLE.md) on how to type props based on existing `propTypes`.
+
+  ```tsx
+  type MyComponentProps = {
+    requiredProp: string;
+    optionalPropWithDefaultValue?: number;
+    optionalProp?: boolean;
+  };
+
+  function MyComponent({
+    requiredProp,
+    optionalPropWithDefaultValue = 42,
+    optionalProp,
+  }: MyComponentProps) {
+    // component's code
+  }
+  ```
+
+<a name="utility-types"></a><a name="1.12"></a>
+
+- [1.12](#utility-types) **Utility Types**: Use types from [TypeScript utility types](https://www.typescriptlang.org/docs/handbook/utility-types.html) and [`type-fest`](https://github.com/sindresorhus/type-fest) when possible.
+
+  ```ts
+  type Foo = {
+    bar: string;
+  };
+
+  // good
+  type ReadOnlyFoo = Readonly<Foo>;
+
+  // bad
+  type ReadOnlyFoo = {
+    readonly [Property in keyof Foo]: Foo[Property];
+  };
```

Comment:
```suggestion
  // BAD
  type ReadOnlyFoo = {
    readonly [Property in keyof Foo]: Foo[Property];
  };

  // GOOD
  type ReadOnlyFoo = Readonly<Foo>;
```

@hayata-suenaga Can we invert the order to keep consistency with other examples?

Reply:
nice catch 👍